### PR TITLE
[4.0] Add publish up/down information to articles list

### PR DESCRIPTION
**Pull request by bembelimen**
Pull Request for Issue #21930 .

### Summary of Changes
Add tooltip information for publish_up/down information + code improvements

### Testing Instructions
Set publish up/down dates, hover the icons in the stage column.

### Expected result
Popover information appear.

### Actual result
No popover information

### CODE REVIEW

File: administrator/language/en-GB/en-GB.com_content.ini [hunk 1]
```diff
@@ -14,6 +14,7 @@ COM_CONTENT_ATTRIBS_FIELDSET_LABEL="Options"
 COM_CONTENT_BATCH_OPTIONS="Batch process the selected articles"
 COM_CONTENT_BATCH_TIP="If a category is selected for move/copy, any actions selected will be applied to the copied or moved articles. Otherwise, all actions are applied to the selected articles."
 COM_CONTENT_CHANGE_ARTICLE="Select or Change article"
+COM_CONTENT_CHANGE_STAGE="Change stage"
```
Comment (infograf768) [inline administrator/language/en-GB/en-GB.com_content.ini]: Alpha order

**Comment by brianteeman** [comment]
I have tested this item :red_circle: unsuccessfully on 8cc3ae8f61483766dd1b952dc981385606fa1a2d<br /><br />i thought it was working great and was about to post thank you etc when i decided to see what happens when you change the state. With this pr when you try to change the state of any article you get an error message

image

<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/21942">issues.joomla.org/tracker/joomla-cms/21942</a>.</sub>

**Comment by bembelimen** [comment]
Thanks @brianteeman

File: layouts/joomla/button/transition-button.php [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+extract($displayData, EXTR_OVERWRITE);
+
+$only_icon = empty($options['transitions']);
+$disabled = !empty($options['disabled']);
+$tip = !empty($options['tip']);
+$id = (int) $options['id'];
+$tipTitle = $options['tip_title'];
+$checkboxName = $options['checkbox_name'];
+?>
+<?php if($only_icon): ?>
```
Comment (Quy) [inline layouts/joomla/button/transition-button.php]: Add space after `if` and before `:`

File: layouts/joomla/button/transition-button.php [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+	<span class="tbody-icon mr-1">
+		<span class="<?php echo $this->escape($icon ?? ''); ?> <?php echo $tip ? 'hasPopover' : '' ?>"
```
Comment (Quy) [inline layouts/joomla/button/transition-button.php]: Add semicolon and repeat below.

File: layouts/joomla/button/transition-button.php [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+			title="<?php echo HTMLHelper::_('tooltipText', Text::_($tipTitle ? : $title), '', 0) ?>"
+			data-content="<?php echo HTMLHelper::_('tooltipText', Text::_($title), '', 0) ?>"
+			data-placement="top"
+		></span>
+	</span>
+	<div class="mr-auto"><?php echo $this->escape($options['stage']); ?></div>
+<?php else: ?>
+	<a class="tbody-icon mr-1 data-state-<?php echo $this->escape($value ?? ''); ?> <?php echo $this->escape(!empty($disabled) ? 'disabled' : null); ?> <?php echo $tip ? 'hasPopover' : '' ?>"
+		<?php if (empty($disabled)): ?>
```
Comment (Quy) [inline layouts/joomla/button/transition-button.php]: Add space before `:`

File: layouts/joomla/button/transition-button.php [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+		<?php
+			$default = [
+				JHtml::_('select.option', '', $this->escape($options['stage'])),
+				JHtml::_('select.option', '-1', '--------', ['disable' => true])
+			];
+
+			$transitions = array_merge($default, $options['transitions']);
+
+			$attribs = [
+				'id'	=> 'transition-select_' . (int) $id,
```
Comment (Quy) [inline layouts/joomla/button/transition-button.php]: Remove extra spaces before `=>` and repeat below.

File: layouts/joomla/button/transition-button.php [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+	<span class="tbody-icon mr-1">
+		<span class="<?php echo $this->escape($icon ?? ''); ?> <?php echo $tip ? 'hasPopover' : ''; ?>"
+			title="<?php echo HTMLHelper::_('tooltipText', Text::_($tipTitle ? : $title), '', 0); ?>"
+			data-content="<?php echo HTMLHelper::_('tooltipText', Text::_($title), '', 0); ?>"
+			data-placement="top"
+		></span>
+	</span>
+	<div class="mr-auto"><?php echo $this->escape($options['stage']); ?></div>
+<?php else: ?>
+	<a class="tbody-icon mr-1 data-state-<?php echo $this->escape($value ?? ''); ?> <?php echo $this->escape(!empty($disabled) ? 'disabled' : null); ?> <?php echo $tip ? 'hasPopover' : '' ?>"
```
Comment (Quy) [inline layouts/joomla/button/transition-button.php]: Add semicolon

File: administrator/components/com_content/tmpl/articles/default.php [hunk 7]
```diff
@@ -200,42 +191,28 @@
...
+													->removeState(1)
+													->removeState(2)
+													->removeState(-2)
+													->addState(ContentComponent::CONDITION_PUBLISHED, '', 'publish', 'COM_CONTENT_CHANGE_STAGE', ['tip_title' => 'JPUBLISHED'])
+													->addState(ContentComponent::CONDITION_UNPUBLISHED, '', 'unpublish', 'COM_CONTENT_CHANGE_STAGE', ['tip_title' => 'JUNPUBLISHED'])
+													->addState(ContentComponent::CONDITION_ARCHIVED, '', 'archive', 'COM_CONTENT_CHANGE_STAGE', ['tip_title' => 'JARCHIVED'])
+													->addState(ContentComponent::CONDITION_TRASHED, '', 'trash', 'COM_CONTENT_CHANGE_STAGE', ['tip_title' => 'JTRASHED'])
+													->setLayout('joomla.button.transition-button')
+													->render($item->stage_condition, $i, $options, $item->publish_up, $item->publish_down);
+
```
Comment (Quy) [inline administrator/components/com_content/tmpl/articles/default.php]: Remove extra blank lines

File: layouts/joomla/button/transition-button.php [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+<?php if ($only_icon) : ?>
+	<span class="tbody-icon mr-1">
+		<span class="<?php echo $this->escape($icon ?? ''); ?> <?php echo $tip ? 'hasPopover' : ''; ?>"
+			title="<?php echo HTMLHelper::_('tooltipText', Text::_($tipTitle ? : $title), '', 0); ?>"
+			data-content="<?php echo HTMLHelper::_('tooltipText', Text::_($title), '', 0); ?>"
+			data-placement="top"
+		></span>
+	</span>
+	<div class="mr-auto"><?php echo $this->escape($options['stage']); ?></div>
+<?php else: ?>
```
Comment (Quy) [inline layouts/joomla/button/transition-button.php]: Add space before `:`

File: layouts/joomla/button/transition-button.php [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+		<?php
+			$default = [
+				JHtml::_('select.option', '', $this->escape($options['stage'])),
+				JHtml::_('select.option', '-1', '--------', ['disable' => true])
+			];
+
+			$transitions = array_merge($default, $options['transitions']);
+
+			$attribs = [
+				'id'		=> 'transition-select_' . (int) $id,
```
Comment (Quy) [inline layouts/joomla/button/transition-button.php]: Remove extra spaces before `=>` and repeat below.

**Comment by brianteeman** [comment]
I have tested this item :white_check_mark: successfully on f179785cda37c0620bdcabb2cf17b2dc15161b50<br /><br />That works now - thanks - one less issue 

Obviously the cs issues need to be fixed but then its good to go<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/21942">issues.joomla.org/tracker/joomla-cms/21942</a>.</sub>

**Comment by ghazal** [comment]
I have tested this item :white_check_mark: successfully on 1a507bc656df8fd3c64550f7dad1b63847bc73c3<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/21942">issues.joomla.org/tracker/joomla-cms/21942</a>.</sub>

**Comment by ghost** [comment]
Ready to Commit after two successful tests.

**Pull request merged**